docs/src/diplomacy/adder_tutorial.md

```diff
@@ -0,0 +1,322 @@
+Similarly, `uFn` maps **incoming upward** parameters into **outgoing upward** ones.
+
+```scala mdoc
+/** node for [[Adder]] (nexus) */
+class AdderNode(dFn: Seq[DownwardParam] => DownwardParam,
+                uFn: Seq[UpwardParam] => UpwardParam)(implicit valName: ValName)
+  extends NexusNode(AdderNodeImp)(dFn, uFn)
+```
+
+There are several other node types, which can be found in [Nodes.scala](https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/diplomacy/Nodes.scala).
+
+### Creating the `LazyModule`s
+
+"Lazy" evaluation refers to deferring the evaluation of an expression until
+it is needed.
+
+Parameter negotiation in Diplomacy is done lazily, after the Diplomacy graph
+is created. Because of this, the concrete hardware that we want to parameterize
+must also be generated lazily. Diplomacy provides the `LazyModule` construct for
+writing lazily-evaluated hardware modules.
+
+Let's use `LazyModule` to define our `Adder`. Notice that the creation of the
+Diplomacy graph itself (in this case, the node), is non-lazy. The desired
+hardware for the module must be written inside `LazyModuleImp`.
+
+For this example, we expect one driver that should be driving two addends of an
+equal bit width into the adder, meaning we expect all of our downward flowing
+parameters to be equivalent. We also expect a sole monitor, so all of our
+upward flowing parameters should also be equivalent. Thus, we can require these
+properties in our `AdderNode` and simply pass on downward the first
+`DownwardParam` we see. Following the same logic, we can pass upward the
+first `UpwardParam` we see.
+
+```scala mdoc
+/** adder DUT (nexus) */
+class Adder(implicit p: Parameters) extends LazyModule {
+  val node = new AdderNode (
+    { case dps: Seq[DownwardParam] =>
+      require(dps.forall(dp => dp.width == dps.head.width))
+      dps.head
+    },
+    { case ups: Seq[UpwardParam] =>
+      require(ups.forall(up => up.width == ups.head.width))
+      ups.head
+    }
+  )
+  lazy val module = new LazyModuleImp(this) {
+    node.out.head._1 := node.in.head._1 + node.in.tail.head._1
+  }
+
+  override lazy val desiredName = "Adder"
+}
+```
+
+Our `AdderDriver` randomly generates addends each cycle. Note that each node
+passes on two identical `width` parameters to be consumed by the `Adder` and
+`AdderMonitor`.
+
+```scala mdoc
+/** driver (source) */
+class AdderDriver(width: Int)(implicit p: Parameters) extends LazyModule {
+  val nodeA = new AdderDriverNode(Seq(DownwardParam(width), DownwardParam(width)))
+  val nodeB = new AdderDriverNode(Seq(DownwardParam(width), DownwardParam(width)))
+
+  lazy val module = new LazyModuleImp(this) {
+    // generate random addends
+    val randomAddendA = FibonacciLFSR.maxPeriod(nodeA.out.head._2.width)  // notice that we are using negotiated
+    val randomAddendB = FibonacciLFSR.maxPeriod(nodeB.out.head._2.width)  // parameters here
```

Comment:
It's a little off that you take `head` here but then `foreach` below. i'd maybe recommend

```
def generateRandomAddend(width: Int) = FibonacciLFSR.maxPeriod(width)

nodeA.out.foreach { case (outputAddend, edgeParam) => outputAddend := generateRandomAddend(edgeParam.width) }
nodeB.out.foreach { case (outputAddend, edgeParam) => outputAddend := generateRandomAddend(edgeParam.width) }
```

Reply:
in your code snippet, wouldn't that generate different addends per node? i.e. `nodeA` would pass two different random numbers into the adder/checker instead of the same random number.

That's why I called the LFSR once with the `head` width, not sure if you know of a better way to do it.